beacon-chain/state/state-native/getters_validator.go

```diff
@@ -413,3 +414,54 @@ func (b *BeaconState) inactivityScoresVal() []uint64 {
 	copy(res, b.inactivityScores)
 	return res
 }
+
+// ActiveBalanceAtIndex returns the active balance for the given validator.
+//
+// Spec definition:
+//
+//	def get_active_balance(state: BeaconState, validator_index: ValidatorIndex) -> Gwei:
+//	    max_effective_balance = get_validator_max_effective_balance(state.validators[validator_index])
+//	    return min(state.balances[validator_index], max_effective_balance)
+func (b *BeaconState) ActiveBalanceAtIndex(i primitives.ValidatorIndex) (uint64, error) {
+	if b.version < version.Electra {
+		return 0, errNotSupported("ActiveBalanceAtIndex", b.version)
+	}
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	if i >= primitives.ValidatorIndex(len(b.validators)) {
+		return 0, errors.Wrapf(consensus_types.ErrOutOfBounds, "validator index %d does not exist", i)
+	}
+	v := b.validators[i]
+
+	return min(b.balances[i], helpers.ValidatorMaxEffectiveBalance(v)), nil
+}
+
+// PendingBalanceToWithdraw returns the sum of all pending withdrawals for the given validator.
+//
+// Spec definition:
+//
+//	def get_pending_balance_to_withdraw(state: BeaconState, validator_index: ValidatorIndex) -> Gwei:
+//	    return sum(
+//	        withdrawal.amount for withdrawal in state.pending_partial_withdrawals if withdrawal.index == validator_index)
+func (b *BeaconState) PendingBalanceToWithdraw(idx primitives.ValidatorIndex) (uint64, error) {
+	if b.version < version.Electra {
+		return 0, errNotSupported("PendingBalanceToWithdraw", b.version)
+	}
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	// TODO: Consider maintaining this value in the state, if it's a potential bottleneck.
```

Comment:
assuming this won't be forgotten lol, hopefully can be discussed at the interop

Reply:
I am going to test it at some point. I implemented it per spec, but the way it is used causes it to be called many times. Maintaining the value on write might be an optimization but I haven't had time to bench it yet